### ISSUE TITLE
[v6] Implement serialization groups

### ISSUE THREAD

**Issue by amourlanne** [TypedProject/tsed#808]
## Informations

I would like to implement serialization groups like Symfony [serializer](https://symfony.com/doc/current/serializer.html#using-serialization-groups-annotations)  service.

### Problem

I have an User class wich is my model for converter service and  my typeorm entity.
```ts
@Entity("users")
class User {
    @Property()
    email: string;

    @Property()
    @Column()
    username: string;

    @Enum(RoleEnum)
    @Column({
        type: 'enum',
        enum: RoleEnum,
        default: RoleEnum.User,
    })
    role: RoleEnum;
}
```
My user object for simple example.
```ts
User {
    email: "admin@tsed.io"
    username: "admin",
    role: "admin"
}
```
I would like my api return (and serialize) my user object like that.
| route | JSON response   | 
|----|---|
| `/users/:username/default` | `{ email: "admin@tsed.io", username: "admin", role: "admin" } `  |
| `/users/:username` | `{ username: "admin" } `  |
|  `/admin/users/:username`|  `{ username: "admin", role: "admin" } `  |

So actually to do that, i have to:

- define 2 other class model for serialization (wich is redundant part of User Class)
- specify those class model in `@ReturnType` decorator for that the converter service use those specific model to serialize / deserialize object.

```ts
class GetUserModel {
    @Property()
    username: string;
}

class GetAdminUserModel {
    @Property()
    username: string;

    @Enum(RoleEnum)
    role: RoleEnum;
}
```

```ts
class UserController {

    constructor(private readonly userService: UserService) {
    }

    @Get("/users/:username/default")
    private async getDefaultUser(@PathParams("username") username: string): Promise<User> {
         return this.userService.getOneByUsername(username);
    }

    @Get("/users/:username")
    @ReturnType({type: GetUserModel})
    private async getUser(@PathParams("username") username: string): Promise<GetUserModel> {
         return this.userService.getOneByUsername(username);
    }

    @Get("/admin/users/:username")
    @ReturnType({type: GetAdminUserModel})
    private async getAdminUser(@PathParams("username") username: string): Promise<GetAdminUserModel> {
         return this.userService.getOneByUsername(username);
    }
```

### Solution: serialization/deserialization groups

The feature is to specify serialization/deserialization groups for model properties.

```ts
@Entity("users")
class User {
    @Property()
    @PrimaryGeneratedColumn()
    email: string;

    @Property({
       groups: ['user', 'admin-user']
    })
    // @Groups(['user', 'admin-user'])
    @Column()
    username: string;

    @Property({
       groups: ['admin-user']
    })
    // @Groups(['admin-user'])
    @Enum(RoleEnum)
    @Column({
        type: 'enum',
        enum: RoleEnum,
        default: RoleEnum.User,
    })
    role: RoleEnum;
}
```
And specify wich serialization group you would like to use in converter service for the returned object.
If group is specified, the converter serialize/deserialize only properties that have this group.
If no group is specified, the converter does what it currently does.

So, my controller is like that.
```ts
class UserController {

    constructor(private readonly userService: UserService) {
    }

    @Get("/users/:username/default")
    private async getDefaultUser(@PathParams("username") username: string): Promise<User> {
         return this.userService.getOneByUsername(username);
    }

    @Get("/users/:username")
    @ReturnType({
        group: 'user'
    })
    private async getUser(@PathParams("username") username: string): Promise<User> {
         return this.userService.getOneByUsername(username);
    }

    @Get("/admin/users/:username")
    @ReturnType({
        group: 'admin-user'
    })
  //  @ReturnObject({
  //      converter: {
  //            group: 'admin-user'
  //      }
  //  })
    private async getAdminUser(@PathParams("username") username: string): Promise<User> {
        return this.userService.getOneByUsername(username);
    }
}
```

And i can specify serialization/deserialization group in converter service like that.

```ts
const user: User = {
    email: "admin@tsed.io",
    username: "admin",
    role: "admin"
}
        
const defaultUserObject = this.converterService.serialize(user, {
    type: User,
})
// console.log(defaultUserObject)
// {
//     email: "admin@tsed.io",
//     username: "admin",
//     role: "admin"
// }
        
const getUserObject = this.converterService.serialize(user, {
    type: User,
    group: 'user'
})
// console.log(getUserObject)
// {
//     username: "admin",
// }

 const getAdminUserObject = this.converterService.serialize(user, {
     type: User,
     group: 'user-admin'
 })
 // console.log(getAdminUserObject)
 // {
 //     username: "admin",
 //     role: "admin"
 // }

const object = {
     email: "admin@tsed.io",
     username: "admin",
     role: "admin"
 }

const defaultUser =  this.converterService.deserialize(user, User);
 // console.log(defaultUserObject)
 // User {
 //     email: "admin@tsed.io",
 //     username: "admin",
 //     role: "admin"
 // }

const getUser =  this.converterService.deserialize(user, User, null, {
     group: 'user'
});
 // console.log(getUser)
 // User {
 //     username: "admin",
 // }

const getUserAdmin =  this.converterService.deserialize(user, User, null, {
     group: 'user-admin'
});
// console.log(getUser)
// User {
//     username: "admin",
//     role: "admin"
// }
```
Use with route decorators.
```ts
class UserController {

  @Post("/")
  @ReturnType({
      group: 'user'
  })
  async postUser( @Group("user-admin") @BodyParams() user: User): Promise<Person> {
    console.log(user); // will be deserialized according the group 'user-admin'.
    // User {
    //     username: "admin",
    //     role: "admin"
    // }

    return user; // will be serialized according the group 'user'.
  }
  // {
  //     username: "admin",
  // }
}
```

### Note
The syntaxe is just an example to introduce the "serialization groups" feature. i'm not attach to the syntaxe but i'm attach to this feature. if you agree to implement this feature, i can try to do it myself.

But for now, what do you think about it ?
(i'm sorry for my bad english)

Alexis Mourlanne

**Comment by Romakita** [comment]
Hello @amourlanne 
This feature will complexify the converterService I think and this is a very specific usecase.

Why don't you use the UseAfter decorator and use a middleware to change the response format ?

```typescript
class Controller {
   @Get("/users/:username/default")
   @UseAfter(GroupMiddleware)                           // <can be wrapped into 
   @StoreSet(GroupMiddleware,  {group: 'user'}) // his own decorator>
    private async getDefaultUser(@PathParams("username") username: string): Promise<User> {
         return this.userService.getOneByUsername(username);
    }
}
```
Create the middleware and his decorator (`tsed generate decorator` - Select Ts.ED middleware and his decorator - cli v1.10.1):
```typescript
import {applyDecorators, StoreSet} from "@tsed/core";
import {Middleware, UseAfter, Context} from "@tsed/common";

export interface IGroupOptions {
  group: string
}

@Middleware()
class GroupMiddleware {
  use(@Context() context: Context) {
      const {group} = context.endpoint.get(GroupMiddleware)
      const data = context.data
       
      switch(group) { 
         case "user":
             /// implement stuff
      }
      
      return data
  } 
}

export function Group(options: IGroupOptions): MethodDecorator {
  return applyDecorators(
   StoreSet(GroupMiddleware, options),
   UseAfter(GroupMiddleware)
  );
}
```

With this example your are free to implement the right mapping :)
See you
Romain

**Comment by Romakita** [comment]
Also you can look on Interceptor feature :)

**Comment by amourlanne** [comment]
Hello,

To introduce group serialization, I took a very simple example. But in reallity, the class diagram is more like [that](https://tsed-swagger.000webhostapp.com/class-diagram.svg) and API is more like [that](https://tsed-swagger.000webhostapp.com/)

I take a little bit more complex example.

## Example
### Entity
We have 4 entities (more than 60 in reallity)

```ts
// @Entity("groups")
export class Group {

    @Property()
    id: string;

    @Property()
    name: string;

    @Property()
    creator?: User;

    @Property()
    publications?: Publication[];

    @Property()
    userGroups?: UserGroup[]
}
```
```ts
// @Entity("publications")
export class Publication {

    @Property()
    id: string;

    @Property()
    author?: User;

    @Property()
    group?: Group;

}
```
```ts
// @Entity("publications")
export class User {

    @Property()
    id: string;

    @Property()
    name: string;

    @Property()
    publications?: Publication[];

    @Property()
    groups?: Group[];

    @Property()
    userGroups?: UserGroup[];
}
```

```ts
// @Entity("user_groups")
// @Unique(['group', 'user'])
export class UserGroup {

    @Property()
    group: Group;

    @Property()
    role: string;

    @Property()
    user: User;
}
```
### Example data
I use this simple data implementation for my example
```ts
import {User} from "./Entities/example/User";
import {Publication} from "./Entities/example/Publication";
import {Group} from "./Entities/example/Group";
import {UserGroup} from "./Entities/example/UserGroup";

const user1: User = {
    id: "user1",
    name: "User 1"
}

const group1: Group = {
    id: "group1",
    name: "Group 1"
}

// user1 create the group 1
user1.groups = [group1]
group1.creator = user1;

// user1 join the group1
const userGroup1: UserGroup = {
    user: user1,
    group: group1,
    role: "admin"
}

user1.userGroups = [userGroup1];
group1.userGroups = [userGroup1];

// user1 publish in group 1
const publication1: Publication = {
    id: "publication1",
    author: user1,
    group: group1
}

user1.publications = [publication1];
group1.publications = [publication1];

// console.log(group1);

// Circular ref like lazy loading ORM relations
// <ref *1> {
//     id: 'group1',
//     creator: {
//         id: 'user1',
//         name: 'user 1',
//         groups: [ [Circular *1] ],
//         userGroups: [ [Object] ],
//         publications: [ [Object] ]
//     },
//     userGroups: [ { user: [Object], group: [Circular *1], role: 'admin' } ],
//     publications: [ { id: 'publication1', author: [Object], group: [Circular *1] } ]

export default {
    user1,
    group1,
    publication1
}
```
### Serialization
And i would like, this serialization:

**getgroups**: `/exemple/groups` return list of all groups serialized like:
```
    [] {
        name,
        creator: {
            id
        },
        userGroups: [] {
            user: {
                name
            }
        }
    }
```
**getGroups**: `/exemple/groups/group1` return the group1 object serialized like:
```
    {
         id,
         name,
         creator:  {
             id,
             name
         },
         publications: [] {
             author: {
                 name
             }
         }
    }
```
But i probably would like other serialization for all other routes.

## Actually
### Custom models

I juste created 8 serialization models for every entity. 4 for getGroups serialization, 4 for getGroup serialization.
#### getGroup

```ts
export class GroupModel {

    @Property()
    id: string;

    @Property()
    name: string;

    @Property()
    creator?: CreatorUserModel /* User */;

    @Property()
    @PropertyType(PublicationsPublicationModel)
    publications?: PublicationsPublicationModel /* Publication[] */;

}
```

```ts
export class AuthorUserModel {

    @Property()
    name: string;

}
```

```ts
export class CreatorUserModel {

    @Property()
    id: string;

    @Property()
    name: string;

}
```

```ts
export class PublicationsPublicationModel {

    @Property()
    author?: AuthorUserModel /* User */;

}
```

#### getGroups

```ts
export class GroupModel {

    @Property()
    name: string;

    @Property()
    creator?: CreatorUserModel /* User */;

    @Property()
    @PropertyType(UserGroupsUserGroupModel)
    userGroups?: UserGroupsUserGroupModel[] /* UserGroup[] */;

}
```

```ts
export class CreatorUserModel {

    @Property()
    id: string;

}
```
```ts
export class UserGroupsUserGroupModel {

    @Property()
    user: UserGroupUserModel /* User */;
}
```

```ts
export class UserGroupUserModel {

    @Property()
    name: string;
}

```
### Controller
And finally i have this controller:
```ts
const group1 = exampleData.group1;
const user1 = exampleData.user1;
const publication1 = exampleData.publication1;

@Controller("/example")
export class ExampleController {

    @Get("/groups")
    @ReturnType({
        type: GetGroupsGroupModel
    })
    private async getGroups() : Promise<GetGroupsGroupModel[]> {
        return [group1];
    }

    // getGroups/GroupModel[] {
    //     name,
    //     creator: getGroups/CreatorUserModel {
    //         id
    //     },
    //     userGroups: getGroups/UserGroupsUserGroupModel[] {
    //         user: getGroups/UserGroupsUserModel {
    //             name
    //         }
    //     }
    // }

    @Get("/groups/group1")
    @ReturnType({
        type: GetGroupGroupModel
    })
    private async getGroup1() : Promise<GetGroupGroupModel> {
        return group1;
    }

    // getGroup/GroupModel {
    //      id,
    //      name,
    //      creator: getGroup/CreatorUserModel {
    //          id,
    //          name
    //      },
    //      publications: getGroup/PublicationsPublicationModel[] {
    //          author: getgroup/AuthorUserModel {
    //              name
    //          }
    //      }
    // }
}
```

## With serialization groups
### Original models

```ts
// @Entity("groups")
export class Group {

    @Property()
    @Groups(['get-group'])
    id: string;

    @Property()
    @Groups(['get-group', 'get-groups'])
    name: string;

    @Property()
    @Groups(['get-group', 'get-groups'])
    creator?: User;

    @Property()
    @Groups(['get-group'])
    publications?: Publication[];

    @Property()
    @Groups(['get-groups'])
    userGroups?: UserGroup[]
}
```
```ts
// @Entity("publications")
export class Publication {

    @Property()
    id: string;

    @Property()
    @Groups(['get-group'])
    author?: User;

    @Property()
    group?: Group;

}
```
```ts
// @Entity("users")
export class User {

    @Property()
    @Groups(['get-group-creator', 'get-groups-creator'])
    id: string;

    @Property()
    @Groups(['get-group-creator', 'get-group-author', 'get-groups-user'])
    name: string;

    @Property()
    publications?: Publication[];

    @Property()
    groups?: Group[];

    @Property()
    userGroups?: UserGroup[];
}
```

```ts
// @Entity("user_groups")
// @Unique(['group', 'user'])
export class UserGroup {

    @Property()
    group: Group;

    @Property()
    role: string;

    @Property()
    @Groups(['get-groups'])
    user: User;
}
```

### Simple Controller

```ts
const group1 = exampleData.group1;
const user1 = exampleData.user1;
const publication1 = exampleData.publication1;

@Controller("/example")
export class ExampleController {

    @Get("/groups")
    @ReturnObject({
        group: ["get-groups", "get-groups-creator", "get-groups-user"]
    })
    private async getGroups() : Promise<Group[]> {
        return [group1];
    }

    // Group:get-groups[] {
    //     name,
    //     creator: User:get-groups-creator {
    //         id
    //     },
    //     userGroups: UserGroup:get-groups[] {
    //         user: User:get-groups-user {
    //             name
    //         }
    //     }
    // }

    @Get("/groups/group1")
    @ReturnObject({
        groups: ["get-group", "get-group-creator", "get-group-author"]
    })
    private async getGroup1() : Promise<Group> {
        return group1;
    }

    // Group:get-group {
    //      id,
    //      name,
    //      creator: User:get-group-creator {
    //          id,
    //          name
    //      },
    //      publications: Publication:get-group[] {
    //          author: User:get-group-author {
    //              name
    //          }
    //      }
    // }
}
```


## Conclusion
It's redundant to create interceptor or custom middleware with big "switch" for all routes and models.
I can create lots of model wich represent a part of original entity model entity, but it is verry redundant and verry complicated. Just for one entity (Group) in only 2 routes, I created 8 Models.

With serialization groups, no need more model.

In my example, I only show groups for serialization, but it also improve security in deserialization model, like submit data wich are a part of an entity.
For example, data to update user. If user send "user.role", this property must not be deserialized because user is not allowed to update is own role. But property is originaly present in user model.

If you are still not convinced of the usefulness of serialization groups, it is not serious, I would use different models.

Alexis

**Comment by chiqui3d** [comment]
I agree with the request, my current solution is the same as what you have done. Creating another class with the necessary properties. 

It would be similar to this:
https://github.com/typestack/class-transformer#using-groups-to-control-excluded-properties

**Comment by Romakita** [comment]
Ok. I see :)
I have a time @amourlanne, let's go.

For the property decorator, I'll take your proposal:

```typescript
class Model {
   @Groups('get-group-creator', 'get-group-author', 'get-groups-user')
   prop: string;
}
```
And for the endpoint decorator:
```typescript
class Controller {
  @Get('/')
  @ReturnType(Model) // don't change this decorator because I work on the Ts.ED v6 (with big improvement for the schema and returned value declaration)
  @UseGroups('group1', 'group2')
  get(){
  }
}
```

**Comment by chiqui3d** [comment]
Perfect, it looks good!

**Comment by amourlanne** [comment]
Have you been able to start the feature ? Otherwise, I can start it now :)

**Comment by Romakita** [comment]
Hello @amourlanne 
You can start the dev, but I'm afraid that you will have to redo your PR after the publication of the new packages @tsed/schema and @tsed/json-mapper.
These two packages will resolve a lot of issues over current JsonSchema and ConverterService. And these packages are independent from the @tsed/common package. 

You can see packages here:
https://github.com/TypedProject/tsed/tree/feat-schema-json-mapper/packages/schema
https://github.com/TypedProject/tsed/tree/feat-schema-json-mapper/packages/json-mapper

The new decorator @Ignore support callback (https://github.com/TypedProject/tsed/blob/feat-schema-json-mapper/packages/schema/src/decorators/common/ignore.ts)
The new @tsed/schema is designed to support groups features and other futures request features.

Maybe the solution could be to start your feature from `feat-schema-json-mapper` branch. Right now, the work is in progress, and the two packages aren't used by @tsed/common. My work over JsonSchema and JsonMapper is done, but  not over OpenSpec some features about operations aren't implement and it's mandatory to stay ISO features with the current framework version.

So tell me what you prefer :)
See you

**Comment by Romakita** [comment]
Hello @amourlanne,

Implement this story on v6 will be more easier rather than in v5. I propose to start the implementation in v6.

In v6, the serialization is managed by @tsed/json-mapper and @tsed/schema. @tsed/schema provide an Ignore decorator which accept a calback to determine when the props must be ignored or not (https://github.com/TypedProject/tsed/blob/alpha/packages/schema/src/decorators/common/ignore.ts#L64)

By using this decorator, we can ignore or keep props in the serialization/deserialization depending the context where the object is serialized.
The Ignore works, but the context declaration is missing (eg: UseGroups). Groups decorator can use the Ignore decorator to implement the expected behavior.

What do you think @amourlanne? If you have any question tell me :)

Romain

**Comment by Romakita** [comment]
# Solution

We will use minimatch module to resolve group configuration.

## Model declaration
```typescript
class Model {
   @Property('!creation') // allow negation by using "!"
   id: string;

   @Groups('group.summary')
   prop1: string;

  @Groups('group.extended')
  prop2: string;

  @Property()
  prop3: string;
}
```

## Usage with Returns decorator

```typescript
class Controller {
  @Get('/')
  @Returns(200, Model).Groups('group.*')  // allow wildcare *
  get(){
     return new Model({...}}
  }
}
```

Expected:

```
{
   id: 'id',
   prop1: 'prop1',
   prop2: 'prop2'
   prop3: 'prop3'
}
```

## Usage with param decorator

```typescript
class Controller {
  @Post('/')
  post(@BodyParams() @Groups('creation', '!group.extended') model: Model){
     console.log(model)
  }
}
```

expected:

```
{
    prop1: 'prop1',
    prop3: 'prop3'
}
```

## Acceptance criteria

- [ ] JsonMapper serialize and deserialize correctly the model according to the group configuration
- [ ] JsonSchema are generated according the group configuration
- [ ] Generated JsonSchema are cached
- [ ] Groups are documented on the website

**Comment by Romakita** [comment]
PR #1133

**Comment by Romakita** [comment]
Hello @amourlanne @mopcweb @chiqui3d 

The groups decorator is available with v6.14.0. See https://tsed.io/docs/model.html#groups

:D

**Comment by amourlanne** [comment]
thanks
